### DISCUSSION TITLE
Checking if a returned promise is complete

### DISCUSSION THREAD

**Discussion by ConnerPhillis** [sebastienros/jint#935] [Q&A]
Hi all,

I just started messing around with the new Promise API and had a quick question. 

in the following example `obj.SimpleTask` is a CLR function returning a task, I have written a custom object converter to swap that to an object. The purpose of the function is to wait for a second and then return a random number, simulating IO work.

```cs
var result = engine.Evaluate("obj.SimpleTask().then(() => 'done')");
var unwrappedResult = result.UnwrapIfPromise();
```

When I run the program, I get the following exception: 
```
System.InvalidOperationException: ''UnwrapIfPromise' called before Promise was settled'
```

When I add a delay that exceeds the duration of the original wait, as in the following, I am able to extract the result, (in this case 'done') from the promise.
```cs
var result = engine.Evaluate("obj.SimpleTask().then(() => 'done')");
await Task.Delay(TimeSpan.FromSeconds(2));
var unwrappedResult = result.UnwrapIfPromise();
```

So far I haven't been able to find a way to check if the promise that has been extracted from the engine has resolved. When I looked in to `UnwrapIfPromise()` I saw that there is a promise state object, but with `PromiseInstance` being an internal class I can't do a type cast to monitor those values.

I'm assuming that I'm probably just missing something, so if any of you have any idea how to check a promise's status let me know.

Thanks in advance!

**Comment by lahma** [comment]
Converted to discussion and pinging @twop for possible insights and ideas.

**Reply by twop** [reply]
Hi @ConnerPhillis good question!

I think the mental model of a task does not cleanly map to Promise in this case.

E.g. when you call `obj.SimpleTask()` I assume you create "ManualPromise" that has explicit resolve/reject methods. And then you add a callback when the task completes and resolve the promise.

Jint doesn't know when the Task completes because it was created in user code.

So before calling `UnwrapIfPromise` you need to make sure that all manual promises were resolved.

E.g. (pseudocode)

```csharp
List<Task> pendingTasks;
engine.Set("SimpleTask", () => {
  var task = StartTask();
  var promise = engine.CreateManaulPromise();
  task.ContinueWith(() => promise.Resolve());
  pendingTasks.Add(task); // <------ keep track of all pending
  return promise;
})

var result = engine.Evaluate("obj.SimpleTask().then(() => 'done')");
await Task.All(pendingTasks);// <---- wait until all are finished
// and now it is safe to unwrap because you KNOW that all tasks were completed
var unwrappedResult = result.UnwrapIfPromise();
```

So conceptually you need to keep track of Tasks you started that map to Promises somewhere outside of Jint.

Hope that helps.

**Reply by ConnerPhillis** [reply]
Hey @twop,

I'm sorry, I have to disagree

From (what I think is) the latest [ECMAScript spec](https://262.ecma-international.org/11.0/#sec-promise-objects)
> A Promise is an object that is used as a placeholder for the eventual results of a deferred (and possibly asynchronous) computation.

Also - here's my code to convert the promise, note that I haven't covered tasks with values yet.

```cs
public bool TryConvert(Engine engine, object value, out JsValue result)
{
	result = JsValue.Undefined;
	if (value is not Task task)
		return false;
	
	var (promise, resolve, reject) = engine.RegisterPromise();
	Task.Run(async () =>
	{
		try
		{
			await task;
			resolve(JsValue.Undefined);
		}
		catch (Exception e)
		{
			reject(JsValue.FromObject(engine, e));
		}
	});
	result = promise;
	return true;
}
```

What I have sent back to jint is a promise object that represents a placeholder for the eventual results of a completed task. I am by no means the best at reading and adhering to specifications, so I recognize there could be flaws in my argument, but I personally feel like I've satisfied the mapping.

Also, I want to touch on the task tracking

>So conceptually you need to keep track of Tasks you started that map to Promises somewhere outside of Jint.

Let's say that what I want out of Jint's `Evaluate` function is a JSON object from an external web API and I want to allow the user to script this depending on the data source they're targeting. The JS code may look something like this:

```js
function fetchApiKey() { ... } 
function executeApiRequest(apiKey) { ... }

const requestResultPromise = fetchApiKey().then(executeApiRequest)

return requestResultPromise
```

In your proposed solution you suggest using `Task.WhenAll`, The problem here would be that by the time `Evaluate` is done we would probably still be waiting on the `fetchApiKey` function. The list of tasks passed to `Task.WhenAll` would not include the `executeApiRequest` task yet, however, so `Task.WhenAll` would complete prematurely.

Additionally, while I can't confirm this without writing test code, when the task yields the second one may not have been added to the list in time, meaning if we checked the list again after `WhenAll` is done the new task representing encapsulating the next promise may not be present.


I noticed [this](https://github.com/sebastienros/jint/blob/1a9cd0ede6df454e9d773a55577b4df8408cc737/Jint/Engine.cs#L329) in the comment for Engine.RegisterPromise, this seems to be what I'm after, but I couldn't find this `ExecuteWithEventLoop` function. If that exists can someone point me to it?

**Comment by twop** [comment]
> I'm sorry, I have to disagree

Happy to elaborate!

The fundamental difference between Task and Promise is that Promise assumes a single threaded EventLoop while Task has a luxury to have an extra feature of multithreading and be more flexible in terms of runtime (no EventLoop). In other words, you HAVE to think through if and how you need to synchronize Tasks, while Promises are inherently sequential via EventLoop. But I totally agree with you that they have more in common than different, but the devil in the details which are very relevant in the context of JS engine implementation.


> In your proposed solution you suggest using Task.WhenAll, The problem here would be that by the time Evaluate is done we would probably still be waiting on the fetchApiKey function. The list of tasks passed to Task.WhenAll would not include the executeApiRequest task yet, however, so Task.WhenAll would complete prematurely.

> Additionally, while I can't confirm this without writing test code, when the task yields the second one may not have been added to the list in time, meaning if we checked the list again after WhenAll is done the new task representing encapsulating the next promise may not be present.

Absolutely spot on!

When you run `resolve(JsValue.Undefined);` it executes JS code that was scheduled with `then` which in turn can produce more promises. So it actually akin to a recursive problem. Here are the steps

1. Create an empty List of pending Taks/Promises
2. Configure engine with the right built-ins (that add to the list) 
3. Run the code with `Evaluate`
4. Check if the pending list empty
5.a. If yes you are done!! and it is safe to "Unwrap" the result
5.b. If not you need to wait for the Pending promises. 
    - Copy the list into a temp var `waitList`
    -  clear the current list
    - await `Task.WhenAll(waitList)`
    - Goto "4"

Note that `resolve(JsValue.Undefined)` triggers JS code execution and should not be called concurrently. Meaning that it is up to you to ensure "singlethreadness" of JS execution.

Btw that is almost exactly how I use Jint Promises in my project.

I think conceptually you can think about it as of mapping Task + .NET runtime scheduling to JS EventLoop.

Does it make it clearer or even more confusing?

**Reply by ConnerPhillis** [reply]
> The fundamental difference between Task and Promise is that Promise assumes a single threaded EventLoop while Task has a luxury to have an extra feature of multithreading and be more flexible in terms of runtime

Thanks for clarifying! I'll try to elaborate a little bit further on my points.

I understand the difference between the Javascript EventLoop and C#'s TPL, but I want to try to set forth another example of what I'm talking about. disclaimer: of this part of JS's event loop I'm only partially familiar, and again can definitely be wrong.

Let's say JS is running through a large period of synchronous work, say it takes about a second to complete. During this work several web request responses make it back to the front end. Those promises get resolved and their values are set even though JS has not yet started working on their continuation activities.

I'm looking at the source code for engine and seeing that there is a private `_eventLoop` variable kept there. Would it be possible to add some property like `HasOutstandingPromises` or something that checks the event loop for promises that have not yet resolved?

**Reply by ConnerPhillis** [reply]
Also, can you elaborate on this?
> Configure engine with the right built-ins (that add to the list)

My concern boils down to this scenario:
1. Task Completes for nested promise which did not queue a task itself, 
  1.1 WhenAll continues
2. non-blocking long runtime code starts running
  2.1. non-blocking, long runtime code continues, eventually queueing a new task containing the resolve for the original
3. WhenAll completes at the same time with another thread
  3.1 non-blocking, long lasting code continues to run on another thread, however no tasks exist on the task list. An attempt is made to unwrap, call fails and exception thrown.
4. non-blocking, long running code continues, task is added to queue post-hoc.

**Reply by twop** [reply]
> Let's say JS is running through a large period of synchronous work, say it takes about a second to complete. During this work several web request responses make it back to the front end. Those promises get resolved and their values are set even though JS has not yet started working on their continuation activities

Not 100% accurate. The completion is being scheduled, it is not immediate. E.g. you cannot ask for the promise state while the 1s loop.

**Reply by ConnerPhillis** [reply]
Sorry - I meant to go back and research that before I made that comment

Either way on the topic of adding a property to check for outstanding promises, I thought of also maybe an event that fires whenever all promises are complete? curious to hear your thoughts.

**Reply by twop** [reply]
> I'm looking at the source code for engine and seeing that there is a private _eventLoop variable kept there. Would it be possible to add some property like HasOutstandingPromises or something that checks the event loop for promises that have not yet resolved?

I had this initially. But then it got really weird with ecma test. E.g `new Promise(() => {})`. In this case `HasOutstandingPromises` will always return `true`. 

Then it might make sense to only track "ManualPromises" (created by the embedder), but then you have full control over creation process (because it was you who created them, right?), thus the property seemed to be redundant.

**Reply by ConnerPhillis** [reply]
Forgive me again, I am unfamiliar with all this,

I ran a test in the browser and saw that the result of that is a promise whose then never executes, would it not be the same with Jint?

**Reply by twop** [reply]
It might be beneficial to rework your initial example a bit to emphasize the "singlethreadness"

```csharp
public bool TryConvert(Engine engine, object value, out JsValue result)
{
	result = JsValue.Undefined;
	if (value is not Task task)
		return false;
	
	var (promise, resolve, reject) = engine.RegisterPromise();
	Task.Run(async () =>
	{
		try
		{
			await task;
                        ScheduleOnMainThread( () => resolve(JsValue.Undefined));
		}
		catch (Exception e)
		{
                        ScheduleOnMainThread( () => reject(JsValue.FromObject(engine, e)));
		}
	});
	result = promise;
	return true;
}
```

where `ScheduleOnMainThread` can be replaced by a dedicated thread assigned to the Engine. In other words, the IO can happen concurrently and multithreaded (like browsers do with fetch) but **completion** of promises has to be scheduled on the same thread.

**Reply by twop** [reply]
> I ran a test in the browser and saw that the result of that is a promise whose then never executes, would it not be the same with Jint?

Exactly the same.

**Reply by twop** [reply]
> My concern boils down to this scenario:

Your concern is totally valid. It touches fundamental point: **can you complete promises concurrently** and the answer is no. At least "no" according to JS spec. In reality you will have to use  `ScheduleOnMainThread` or similar construct to `complete` promises.

With this constraint in place you won't have issues described in the scenario.

**Reply by ConnerPhillis** [reply]
I think I'll try to get to the essence of my argument here,

to me, asking every caller who wants to invoke an asynchronous function to create some control construct around engine to ensure the single-threadedness of javascript is conserved sounds like an unnecessary and heavy burden to place on the callers.

Personally, I feel like it would be far easier to place a lock inside the resolve / reject / execute functions, that way the threads that are trying to run the resolutions aren't going to step on one another. (caveat being there may be some deterministic scheduling process in the promise specification that C# may have difficulty implementing) Also, with permission from the maintainers, I am willing to work on that, I'm not asking anyone else to do it.

**Reply by ConnerPhillis** [reply]
Additionally, per the `new Promise(() => {})` example, I would consider including that as an outstanding promise. 

To me having the functionality that allows you to verify that a script is _completely_ done versus done executing all of the synchronous code is an absolute must, and if the person writing a script shoots themselves in the foot by leaving in redundant code like that then that's not something a library maintainer should have to design against.

I also want to quickly emphasize that in no way is any of this criticism of your code. This feature is far and above what I would be able to build at this point in my life, and your knowledge of javascript is clearly bounds beyond mine. I just feel like there could be a few more improvements around the C# interop. You have my utmost respect for the work that you've done.

**Reply by twop** [reply]
> Personally, I feel like it would be far easier to place a lock inside the resolve / reject / execute functions, that way the threads that are trying to run the resolutions aren't going to step on one another. (caveat being there may be some deterministic scheduling process in the promise specification that C# may have difficulty implementing) Also, with permission from the maintainers, I am willing to work on that, I'm not asking anyone else to do it.

That is exactly how I started implementing the feature but quickly changed my direction. Now I feel strongly that the burden to sync Task should be done outside of Jint. 

There are several reasons for that.
1. The way it was done initially implied overhead regardless if you use promises or not. It can be mitigated by something like `ExecuteAsync` though. When I tried to do just that I quickly stuck at reworking all the ecma262 tests to convert to async execution.
2. There has to be a way to create a promise when you don't have an underlying Task. E.g. I need to have a manual way to resolve Promise.
3. Adding sync code to Jint makes it more heavyweight both in terms of maintenance and code. I use it for mobile App and size is a concern for me.

It is possible that I'm totally wrong there though.

**Reply by twop** [reply]
If I may suggest a path forward for this conversation:

1. Try to experience all the pain of synchronizing yourself as we discussed above and make your API work
2. Try to see if that code can be extracted into a reusable chunk. If yes then it is possible to either bundle that code into Jint itself (and make optional) or even create a new package.
3. If there is no easy way to make it "library" like or it feels awkward then we have a proof that the current API is not good enough

If you would be willing to pursue this route then it would be awesome if you post here gists/snippets of your code, and we can brainstorm further together :)

**Reply by ConnerPhillis** [reply]
Here's the completed sample, this was a quick job so if there are errors please let me know.

```cs
public class SimpleDemoObjects
{
	public async Task<int> SimpleTask()
	{
		await Task.Delay(TimeSpan.FromSeconds(1));
		return new Random((int)DateTime.Now.Ticks).Next();
	}
	public int Fibonacci(int n)
	{
		if (n <= 2)
			return 1;
		return Fibonacci(n - 1) + Fibonacci(n - 2);
	}
}
	
class Program
{
	static async Task Main(string[] args)
	{
		var semaphore = new SemaphoreSlim(1);
		
		var taskList = new List<Task>();
		var converter = new Converter(
			async (action) =>
			{
				/*
				 * I understand this does not actually run on the same thread,
				 * but I believe that this is a close enough parallel
				 */
				try
				{
					await semaphore.WaitAsync();
					action();
				}
				finally
				{
					semaphore.Release();
				}
				
			}, (task) => { taskList.Add(task); });

		var engine = new Engine(setup => { setup.AddObjectConverter(converter); });
		engine.SetValue("obj", new SimpleDemoObjects());

		JsValue result;
		lock (ThreadLock)
		{
			result = engine.Evaluate(
				"obj.SimpleTask()"
				+ ".then(v => { obj.Fibonacci(34) })"
				+ ".then(() => { obj.SimpleTask(); obj.SimpleTask(); return obj.SimpleTask(); } )"
				+ ".then(() => obj.SimpleTask())"
				+ ".then(() => obj.Fibonacci(12))");
		}
		while (taskList.Count > 0)
		{
			var taskArray = taskList.ToArray();
			taskList.Clear();
			await Task.WhenAll(taskArray);
		}
		
		var unwrappedResult = result.UnwrapIfPromise();
		Console.WriteLine(unwrappedResult.ToString());
	}
}

public class Converter : IObjectConverter
{
	private Func<Action, Task> _runOnMainThread;
	private Action<Task> _onTaskAdded;
	public Converter(Func<Action, Task> runOnMainThread, Action<Task> onTaskAdded)
	{
		_runOnMainThread = runOnMainThread;
		_onTaskAdded = onTaskAdded;
	}
	
	public bool TryConvert(Engine engine, object value, out JsValue result)
	{
		result = JsValue.Undefined;
		if (value is not Task task)
			return false;
		var (promise, resolve, reject) = engine.RegisterPromise();
		var newTask = Task.Run(
			async () =>
			{
				try
				{
					await task;
					await _runOnMainThread(() => resolve(JsValue.Undefined));
				}
				catch (Exception e)
				{
					await _runOnMainThread(() => reject(JsValue.FromObject(engine, e)));
				}
			});
		
		_onTaskAdded(newTask);
		result = promise;
		return true;
	}
}
```

The two biggest pain points that I see here are:
- This is a **lot** of code to force every user who wants to use async promises in Jint
- The converter that I've written will have to have a second converter to handle tasks with values, or I'll have to write some reflection-based evaluator, again not easy for callers

**Reply by twop** [reply]
Hey @ConnerPhillis  thanks for sketching that out. After I slept on our points I think it might make sense to try to add an extension to Jint (either bake in or as 3rd party package) that automates the hard parts. 

I might be wrong but I think Jint can map the values of tasks automatically too.

I don't have time today but I might in the next day or two to tinker how it might look like

**Reply by ConnerPhillis** [reply]
Great! happy to be a sounding board for any ideas. Thanks!

**Reply by twop** [reply]
Sorry for being slow on this, life happens :) but I haven't forgotten!

**Reply by twop** [reply]
I think I can make a sketch of the following API

```csharp 
public async void Example()
{
    var (engine, ctx) = EngineAsyncExtension.PrepareWithAsyncContext();
    
    engine.SetValue("obj", new SimpleDemoObjects());

    var res = await engine.EvaluateAsync(ctx, "obj.SimpleTask()"
                                              + ".then(v => { obj.Fibonacci(34) })"
                                              + ".then(() => { obj.SimpleTask(); obj.SimpleTask(); return obj.SimpleTask(); } )"
                                              + ".then(() => obj.SimpleTask())"
                                              + ".then(() => obj.Fibonacci(12))")
}
```
basically `PrepareWithAsyncContext()` will initialize Engine with the right converter for Tasks and `ctx` will keep track of the outstanding tasks.

Note that `EvaluateAsync` is an extension method that internally does the recursive execution magic we were talking about and returns a `Task` created from `TaskCompletionSource`.

Note2: I think I can make this API work purely as a public extension for the Engine. If the API is good and I can make it work then we will see where it should live.

Note3: `engine.EvaluateAsync` is not thread safe. That is, don't call `EvaluateAsync` until the previous `EvaluateAsync` call hasn't finished. BUT the tasks that were produced by `SimpleTask()` are thread safe, e.g. you don't need to worry about sync context in your C# methods.


I assume this API would be acceptable for you? @ConnerPhillis

**Reply by twop** [reply]
Ok I got it working 

```csharp
    public static class AsyncEngine
    {
        public class AsyncContext
        {
            internal readonly SemaphoreSlim EngineSemaphore = new(1);
            internal uint PendingTasksCount { get; set; }
            internal Action OnLastTaskFinished { get; set; }
        }

        internal class TaskConverter : IObjectConverter
        {
            private readonly AsyncContext _ctx;

            public TaskConverter(AsyncContext ctx)
            {
                _ctx = ctx;
            }

            public bool TryConvert(Engine engine, object value, out JsValue result)
            {
                result = JsValue.Undefined;
                if (value is not Task task)
                    return false;

                var (promise, resolve, reject) = engine.RegisterPromise();

                async Task SettleOnEngineThread(Action action)
                {
                    var semaphore = _ctx.EngineSemaphore;
                    try
                    {
                        await semaphore.WaitAsync();
                        action();
                    }
                    finally
                    {
                        _ctx.PendingTasksCount--;
                        var finishExecution = _ctx.PendingTasksCount == 0;
                        semaphore.Release();
                        if (finishExecution)
                        {
                            // it is safe to set result outside semaphore because
                            // this task is the last possible race condition

                            _ctx.OnLastTaskFinished();
                        }
                    }
                }

                // Try convert will ALWAYS happen on the same thread as JS execution
                // Thus it is safe to increment it here without the lock
                _ctx.PendingTasksCount++;

                Task.Run(
                    async () =>
                    {
                        try
                        {
                            await task;
                            await SettleOnEngineThread(() => resolve(JsValue.Undefined));
                        }
                        catch (Exception e)
                        {
                            await SettleOnEngineThread(() => reject(JsValue.FromObject(engine, e)));
                        }
                    });

                result = promise;
                return true;
            }
        }

        public static (Engine, AsyncContext) PrepareWithAsyncContext()
        {
            var ctx = new AsyncContext();
            var engine = new Engine(setup => { setup.AddObjectConverter(new TaskConverter(ctx)); });
            return (engine, ctx);
        }

        public static async Task<JsValue> EvaluateAsync(this Engine engine, AsyncContext ctx, string code)
        {
            await ctx.EngineSemaphore.WaitAsync();

            // TODO it should probably be wrapped in assert instead or something
            // it should always be zero at the start of the execution
            // e.g. "assert(ctx.PendingTasksCount, 0);
            ctx.PendingTasksCount = 0;

            var completionSource = new TaskCompletionSource<JsValue>();

            ctx.OnLastTaskFinished =
                () => completionSource.SetResult(engine.GetCompletionValue().UnwrapIfPromise());

            bool finishExecution;

            try
            {
                engine.Execute(code);

                // means that there were no tasks created during execution
                finishExecution = ctx.PendingTasksCount == 0;
            }
            finally
            {
                ctx.EngineSemaphore.Release();
            }


            if (finishExecution)
            {
                return engine.GetCompletionValue().UnwrapIfPromise();
            }

            return await completionSource.Task;
        }
    }

    public class ApiExperiment
    {
        private readonly ITestOutputHelper _testOutputHelper;

        public ApiExperiment(ITestOutputHelper testOutputHelper)
        {
            _testOutputHelper = testOutputHelper;
        }

        public class SimpleDemoObjects
        {
            private Action onStart;
            private Action onFinish;

            public SimpleDemoObjects(Action onStart, Action onFinish)
            {
                this.onStart = onStart;
                this.onFinish = onFinish;
            }

            public async Task<int> SimpleTask()
            {
                onStart();
                //Console.WriteLine("Delay start {0}", DateTime.Now);
                await Task.Delay(TimeSpan.FromSeconds(2));
                // Console.WriteLine("Delay end {0}", DateTime.Now);
                onFinish();
                return new Random((int)DateTime.Now.Ticks).Next();
            }

            public int Fibonacci(int n)
            {
                if (n <= 2)
                    return 1;
                return Fibonacci(n - 1) + Fibonacci(n - 2);
            }
        }

        [Fact(Timeout = 10000)]
        public async void Example()
        {
            var (engine, ctx) = AsyncEngine.PrepareWithAsyncContext();

            engine.SetValue("obj",
                new SimpleDemoObjects(() => _testOutputHelper.WriteLine("Delay start {0}", DateTime.Now),
                    () => _testOutputHelper.WriteLine("Delay end {0}", DateTime.Now)));

            var res = await engine.EvaluateAsync(ctx, "obj.SimpleTask()"
                                                      + ".then(v => { obj.Fibonacci(34) })"
                                                      + ".then(() => { obj.SimpleTask(); obj.SimpleTask(); return obj.SimpleTask(); } )"
                                                      + ".then(() => obj.SimpleTask())"
                                                      + ".then(() => obj.Fibonacci(12))");

            _testOutputHelper.WriteLine("end of execution {0}, res = {1}", DateTime.Now, res);
        }
    }
```

Note that  `ApiExperiment` is a part of a test suite (I debugged as a test).

P.S. as a side benefit `EvaluateAsync` is now thread safe too. Meaning that you can try to run several EvaluateAsync at the same time but only one will be executed at any given point, it is using the same semaphore to access engine.

**Reply by twop** [reply]
@ConnerPhillis  @lahma @sebastienros this is a PoC for supporting mapping of Task -> Promise. Note that it is done completely out of Jint internals, meaning that it can be even a separate package. I personally don't need Tasks (yet) but something along these lines might be valuable for the community.

Note that the actual API details is up to debate, just wanted to sketch out what I had in mind

**Reply by twop** [reply]
as a gist https://gist.github.com/twop/8d7b849d0105c779eed87bff8ae59722

**Reply by twop** [reply]
Btw here is the output
```
Delay start 08/01/2021 10:45:15
Delay end 08/01/2021 10:45:17
Delay start 08/01/2021 10:45:17
Delay start 08/01/2021 10:45:17
Delay start 08/01/2021 10:45:17
Delay end 08/01/2021 10:45:19
Delay end 08/01/2021 10:45:19
Delay end 08/01/2021 10:45:19
Delay start 08/01/2021 10:45:19
Delay end 08/01/2021 10:45:21
end of execution 08/01/2021 10:45:21, res = 144
```

not that there are three starts that happen at the same time which corresponds to: 
`then(() => { obj.SimpleTask(); obj.SimpleTask(); return obj.SimpleTask(); } )`. 

E.g. Tasks are executing in parallel

**Reply by twop** [reply]
and here is a bit more explicit example:

code (note `Promise.all`):

```js
obj.SimpleTask(0)
    .then(v => {
        obj.Fibonacci(34)
    })
    .then(() => Promise.all([obj.SimpleTask(1), obj.SimpleTask(2), obj.SimpleTask(3)]))
    .then(() => obj.SimpleTask(4))
    .then(() => obj.Fibonacci(12))
```

output
```
start 0 at 08/01/2021 11:08:44
end 0 at 08/01/2021 11:08:46
start 1 at 08/01/2021 11:08:46
start 2 at 08/01/2021 11:08:46
start 3 at 08/01/2021 11:08:46
end 1 at 08/01/2021 11:08:48
end 3 at 08/01/2021 11:08:48
end 2 at 08/01/2021 11:08:48
start 4 at 08/01/2021 11:08:48
end 4 at 08/01/2021 11:08:50
end of execution 08/01/2021 11:08:50, res = 144
```

Note that 1, 2,3 started at the same time (in this order) and ended at the same time unlocking the 4th one, but in different order 1,3,2.

**Reply by lahma** [reply]
Great work! Seems quite simple, yet powerful. Some notes about tighter integration (out-of-the-box experience) based on my limited understanding:

* if `AllowClr` is true, we could just add the Task converter
* `EvaluateAsync` could do something like `if(_asyncContext is not null) throw new InvalidOperationException("concurrent calls not allowed")` or similar mechanism to block, so initializing and disposing the context per EvaluateAsync call?
   * the semaphore might get confusing as engine is not thread safe and then this single operation would be
* Would it make sense to expose public API as `ValueTask<JsValue>`?

This is quite integral to interop story so I'd think it would be nice to have it as part of the normal distribution with remark of unstable API or something similar.

**Reply by ConnerPhillis** [reply]
Hey, @twop,

Sorry for not getting back to you earlier, but that's quite impressive! As of right now I can't think of anything that I could need that would not be satisfied with the code that you've provided.

@lahma 
To your second point - `SemaphoreSlim` exposes non-async wait methods, maybe `Evaluate` could be modified to use the same semaphore but calling `Wait` instead, making Jint thread safe. Not sure of the larger implications of this, just wanted to toss an idea in to the hat.

**Reply by lahma** [reply]
> To your second point - `SemaphoreSlim` exposes non-async wait methods, maybe `Evaluate` could be modified to use the same semaphore but calling `Wait` instead, making Jint thread safe. Not sure of the larger implications of this, just wanted to toss an idea in to the hat.

Problem with that is that I've seen people calling `engine.Evalute` inside the own c# code running inside the engine so I would presume that kind code would deadlock as it's a nested call. This is still thread-safe as we expect end user to call single-threadedly and thus no need for locks.

**Reply by twop** [reply]
> Great work! Seems quite simple, yet powerful. 

^_^ 

> Some notes about tighter integration (out-of-the-box experience) based on my limited understanding: if AllowClr is true, we could just add the Task converter
EvaluateAsync could do something like if(_asyncContext is not null) throw new InvalidOperationException("concurrent calls not allowed") or similar mechanism to block, so initializing and disposing the context per EvaluateAsync call?

Unfortunately, the way the code is structured you cannot do it right now. Converter modifies state of `AsyncContext`, meaning that context **has to be created before converter**, thus there is a need for this awkward `PrepareWithAsyncContext`. Note that this pain would go away if you can add and remove converters after Engine creation. Then the entire logic can be moved into just `EvaluateAsync`.

> the semaphore might get confusing as engine is not thread safe and then this single operation would be

Semaphore is needed to make sure Tasks report their results in a queue vs concurrently. Or maybe I misunderstood your question? Note that thread safety of `EvaluateAsync` is just a side benefit of that, we can just omit that from the docs/notes altogether, people should not call `EvaluateAsync` concurrently.



> Would it make sense to expose public API as ValueTask<JsValue>?

Totally, I just modified the original example which used Tasks.

> This is quite integral to interop story so I'd think it would be nice to have it as part of the normal distribution with remark of unstable API or something similar.

If it is possible I would prefer it to be an add-on. Ideally even something like a Jint optional package `JintTask2Promise`.


@ConnerPhillis I would try to avoid using any threadsafety mechanisms if they are not needed. They incur performance and maintenance costs. Before .NET6 tree-shaking is a concern, shipping code that you don't need is not great on constrained environments.

**Reply by lahma** [reply]
> > Some notes about tighter integration (out-of-the-box experience) based on my limited understanding: if AllowClr is true, we could just add the Task converter
> > EvaluateAsync could do something like if(_asyncContext is not null) throw new InvalidOperationException("concurrent calls not allowed") or similar mechanism to block, so initializing and disposing the context per EvaluateAsync call?
> 
> Unfortunately, the way the code is structured you cannot do it right now. Converter modifies state of `AsyncContext`, meaning that context **has to be created before converter**, thus there is a need for this awkward `PrepareWithAsyncContext`. Note that this pain would go away if you can add and remove converters after Engine creation. Then the entire logic can be moved into just `EvaluateAsync`.

I was thinking that `AsyncContext` could have been a field for `Engine` and engine initialization (where options are applied) could check if `AllowClr` and then instantiate the async context and hook the converter (I didn't see the dependency converter has before there, but workable). So engine could instantiate at construction and configure as required, there's all the possibilities when Engine knows about the feature.

> > the semaphore might get confusing as engine is not thread safe and then this single operation would be
> 
> Semaphore is needed to make sure Tasks report their results in a queue vs concurrently. Or maybe I misunderstood your question? Note that thread safety of `EvaluateAsync` is just a side benefit of that, we can just omit that from the docs/notes altogether, people should not call `EvaluateAsync` concurrently.

I understood the semaphore more like being engine lock, but it's context lock, OK.

> > Would it make sense to expose public API as ValueTask?
> 
> Totally, I just modified the original example which used Tasks.

👍🏻 

> 
> > This is quite integral to interop story so I'd think it would be nice to have it as part of the normal distribution with remark of unstable API or something similar.
> 
> If it is possible I would prefer it to be an add-on. Ideally even something like a Jint optional package `JintTask2Promise`.

I still feel that as this is not in it's current form dependent of libraries or special constructs, and being quite small, it could fit the engine as service too.

**Reply by twop** [reply]
> I was thinking that AsyncContext could have been a field for Engine and engine initialization (where options are applied) could check if AllowClr and then instantiate the async context and hook the converter (I didn't see the dependency converter has before there, but workable). So engine could instantiate at construction and configure as required, there's all the possibilities when Engine knows about the feature.

That is a good point about the engine knowing about AsyncContext. I think if this is a built-in feature then maybe dedicate a flag for this. Or have some other 1st class API support for this use case.

Regardless I would love if the Task interop is purely opt-in, and promises can work without it (like I use them in my project).

I don't feel that I have enough stake in the subject to make a call about API surface area, but I would love to be a part of the brainstorming process.

**Comment by twop** [comment]
> I noticed this in the comment for Engine.RegisterPromise, this seems to be what I'm after, but I couldn't find this ExecuteWithEventLoop function. If that exists can someone point me to it?

Good find! It used to be two functions `Execute` `ExecuteWithEventLoop` but now `Execute` combines both, thus the comment is outdated :( Sorry for confusion

**Reply by ConnerPhillis** [reply]
No problem, that makes perfect sense!
